service/matching/ack_manager.go

```diff
@@ -112,29 +128,43 @@
 	}
 }
 
-func (m *ackManager) completeTask(taskID int64) (ackLevel int64) {
+func (m *ackManager) completeTask(taskID int64) int64 {
 	m.Lock()
 	defer m.Unlock()
-	if completed, ok := m.outstandingTasks[taskID]; ok && !completed {
-		m.outstandingTasks[taskID] = true
-		m.backlogCounter.Dec()
+	macked, found := m.outstandingTasks.Get(taskID)
+	acked := macked.(bool)
+	if !found || acked {
+		// don't adjust ack level if nothing has changed
+		return m.ackLevel
 	}
 
 	// TODO the ack level management should be done by a dedicated coroutine
 	//  this is only a temporarily solution
+	m.outstandingTasks.Put(taskID, true)
+	m.backlogCounter.Dec()
 
-	taskIDs := maps.Keys(m.outstandingTasks)
-	util.SortSlice(taskIDs)
+	min, _ := m.outstandingTasks.Min()
+	if taskID != min {
+		return m.ackLevel
+	}
 
-	// Update ackLevel
-	for _, taskID := range taskIDs {
-		if acked := m.outstandingTasks[taskID]; acked {
-			m.ackLevel = taskID
-			delete(m.outstandingTasks, taskID)
-		} else {
-			return m.ackLevel
+	// We've acked the minimum task, so should adjust the ack level as far as we can
```

Comment:
wouldn't it be simpler to write the whole thing (starting from line 134) like this:

```
for {
	min, acked := m.outstandingTasks.Min()
	if min == nil || !acked.(bool) {
		return m.ackLevel
	}
	m.outstandingTasks.Remove(min)
}
```

note that we currently never complete tasks out of order

Reply:
You're right, I overcomplicated this